src/util/side_metadata/helpers.rs

```diff
@@ -0,0 +1,210 @@
+use crate::util::{constants, memory, Address};
+use memory::dzmmap;
+use std::io::{Error, Result};
+
+use super::global::METADATA_SINGLETON;
+use super::SideMetadataID;
+
+#[cfg(target_pointer_width = "32")]
+pub(super) const METADATA_BASE_ADDRESS: Address = unsafe { Address::from_usize(0) };
+#[cfg(target_pointer_width = "64")]
+pub(super) const METADATA_BASE_ADDRESS: Address =
+    unsafe { Address::from_usize(0x0000_0600_0000_0000) };
+
+#[cfg(target_pointer_width = "32")]
+pub(super) const MAX_HEAP_SIZE_LOG: usize = 32;
+// FIXME: This must be updated if the heap layout changes
+#[cfg(target_pointer_width = "64")]
+pub(super) const MAX_HEAP_SIZE_LOG: usize = 48;
+
+pub(super) const MAX_METADATA_BITS: usize = constants::BITS_IN_WORD;
+// const SPACE_PER_META_BIT: usize = 2 << (MAX_HEAP_SIZE_LOG - constants::LOG_BITS_IN_WORD);
+pub(super) const META_SPACE_PAGE_SIZE: usize = constants::BYTES_IN_PAGE;
+pub(super) const META_SPACE_PAGE_SIZE_LOG: usize = constants::LOG_BYTES_IN_PAGE as usize;
+
+#[inline(always)]
+pub(super) fn address_to_meta_address(addr: Address, metadata_id: SideMetadataID) -> Address {
+    let bits_num_log = METADATA_SINGLETON.meta_bits_num_log_vec[metadata_id.as_usize()] as i32;
+    // right shifts for `align` times, then
+    // if bits_num_log < 3, right shift a few more times to cover multi objects per metadata byte
+    // if bits_num_log = 3, metadata byte per object is 1
+    // for > 3, left shift, because more than 1 byte per object is required
+    let rshift = (constants::LOG_BITS_IN_BYTE as i32) - bits_num_log;
+    let offset = if rshift >= 0 {
+        addr.as_usize()
+            .checked_shr(METADATA_SINGLETON.align[metadata_id.as_usize()] as u32)
+            .unwrap()
+            .checked_shr(rshift as u32)
+            .unwrap()
+    } else {
+        addr.as_usize()
+            .checked_shr(METADATA_SINGLETON.align[metadata_id.as_usize()] as u32)
+            .unwrap()
+            .checked_shl(-rshift as u32)
+            .unwrap()
+    };
+
+    METADATA_SINGLETON.meta_base_addr_vec[metadata_id.as_usize()] + offset
+}
+
+// Gets the related meta address and clears the low order bits
+pub(super) fn address_to_meta_page_address(
+    data_addr: Address,
+    metadata_id: SideMetadataID,
+) -> Address {
+    let meta_addr = address_to_meta_address(data_addr, metadata_id);
+    unsafe {
+        Address::from_usize((meta_addr >> META_SPACE_PAGE_SIZE_LOG) << META_SPACE_PAGE_SIZE_LOG)
+    }
+}
+
+// Checks whether the meta page containing this address is already mapped.
+//
+// Returns Err if the address is not mappable by mmtk,
+// and Ok(is_mapped?) otherwise.
+//
+// NOTE: using incorrect (e.g. not properly aligned) page_addr is undefined behavior.
+pub(super) fn meta_page_is_mapped(page_addr: Address) -> Result<bool> {
```

Comment:
1. Right, the limitation is that the current implementation needs the addr to be page-aligned, otherwise the mmap documentation says it may not work properly. We can still handle this by masking the addr and getting a page address (e.g. `is_mapped(addr, size)` calls `is_page_mapped((addr >> 12) << 12, size + (addr - (addr >> 12) << 12))`.

I can move it to util::memory and implement both these functions, if it's useful there.

2. The main goal was to check whether an address is mapped or not. I agree that it's confusing, but I don't think is_mappable improves it.

Reply:
1. You can use the functions in https://github.com/mmtk/mmtk-core/blob/master/src/util/conversions.rs for rounding.
2. OK. It is possible to return an enum for different states (mapped, available, not available). 

A separate issue with this function is that even if we can map it, and say the memory is available, but we unmap it before return. There is no guarantee that next time we try to map the same memory area, it is still available. Maybe once we map the memory, we should not unmap it?

Reply:
Considering that the memory area is exclusive to side metadata, if it's available now, it will be available in the future.

We can't remove the unmapping, because if any step in the mapping fails, we don't know which pages to unmap then.